src/tests/rust_guests/simpleguest/src/main.rs

```diff
@@ -664,1004 +367,269 @@ fn execute_on_heap(_function_call: &FunctionCall) -> Result<Vec<u8>> {
-    ) {
-        let base = base as usize as *mut u8;
-        let len = len as usize;
+#[guest_function("ExecMappedBuffer")]
+fn exec_mapped_buffer(base: u64, len: u64) -> bool {
+    let base = base as usize as *mut u8;
+    let len = len as usize;
 
-        unsafe {
-            hyperlight_guest_bin::paging::map_region(base as _, base as _, len as u64 + 4096)
-        };
+    unsafe { hyperlight_guest_bin::paging::map_region(base as _, base as _, len as u64 + 4096) };
 
-        let data = unsafe { core::slice::from_raw_parts(base, len) };
+    let data = unsafe { core::slice::from_raw_parts(base, len) };
 
-        // Should be safe as long as data is something like a NOOP followed by a RET
-        let func: fn() = unsafe { core::mem::transmute(data.as_ptr()) };
-        func();
+    // Should be safe as long as data is something like a NOOP followed by a RET
+    let func: fn() = unsafe { core::mem::transmute(data.as_ptr()) };
+    func();
 
-        Ok(get_flatbuffer_result(true))
-    } else {
-        Err(HyperlightGuestError::new(
-            ErrorCode::GuestFunctionParameterTypeMismatch,
-            "Invalid parameters passed to read_mapped_buffer".to_string(),
-        ))
-    }
+    true
 }
 
-fn call_host_expect_error(function_call: &FunctionCall) -> Result<Vec<u8>> {
-    if let ParameterValue::String(hostfuncname) =
-        function_call.parameters.clone().unwrap()[0].clone()
-    {
-        let res = call_host_function::<i32>(&hostfuncname, None, ReturnType::Int);
-
-        match res {
-            Ok(_) => Err(HyperlightGuestError::new(
-                ErrorCode::GuestError,
-                "Expected host function to fail, but it succeeded".to_string(),
-            )),
-            Err(e) => {
-                assert_eq!(e.kind, ErrorCode::HostFunctionError);
-                assert_eq!(
-                    e.message,
-                    format!("HostFunction {} was not found", hostfuncname)
-                );
-                Ok(get_flatbuffer_result(()))
-            }
-        }
-    } else {
-        Err(HyperlightGuestError::new(
-            ErrorCode::GuestFunctionParameterTypeMismatch,
-            "Invalid parameters passed to call_host_expect_error".to_string(),
-        ))
-    }
+#[guest_function("CallHostExpectError")]
+fn call_host_expect_error(hostfuncname: String) -> Result<()> {
+    let res = call_host_function::<i32>(&hostfuncname, None, ReturnType::Int);
+
+    let Err(e) = res else {
+        return Err(HyperlightGuestError::new(
+            ErrorCode::GuestError,
+            "Expected host function to fail, but it succeeded".to_string(),
+        ));
+    };
+
+    assert_eq!(e.kind, ErrorCode::HostFunctionError);
+    assert_eq!(
+        e.message,
+        format!("HostFunction {hostfuncname} was not found")
+    );
+    Ok(())
 }
 
 #[no_mangle]
 #[instrument(skip_all, parent = Span::current(), level= "Trace")]
 pub extern "C" fn hyperlight_main() {
```

Comment:
We should update the read me with this new way to register functions. If there is there any reason to register functions manually, we should make a note of it.  

Do we want to keep a few basic manually registrations around as test cases?

Reply:
Yea i think so, I believe we are mostly covered since we are generating the same code, but want to make sure that can still use the old way and we don't introduce subtle bugs in the two different ways of registering functions.